### DISCUSSION TITLE
[question] How to modify the LOGO and webpage theme?

### DISCUSSION THREAD

**Discussion by muzihuaner** [searxng/searxng#974] [Q&A] [answered]
How to modify the LOGO and webpage theme? How to enter the Docker environment?

**Comment by hirad1986** [comment]
In case you're still wondering, You can enter Docker container using :
`docker exec -ti searxng /bin/sh` (rename searxng to whatever the name of your container is)
If you're using the simple theme, you can find the logo img in /usr/local/searxng/searx/static/themes/simple/img/searxng.png
You can mount the file as a volume and give the address of where you keep your custom logo. for example:
`/srv/searxng/searxng-custom.png:/usr/local/searxng/searx/static/themes/simple/img/searxng.png`

**Reply by m0jon** [reply]
I have my custom logo stored in /usr/local/searxng-docker on my server. When in run /usr/local/searxng-docker/searxng.png:/usr/local/searxng/searx/static/themes/simple/img/searxng.png it comes back with not found. Do i need my custom logo within the container before i mount it as a volume or is there a way to link it from outside the container

**Reply by hirad1986** [reply]
It shouldn't be necessary. Because that's the location of the original searxng.png. So you're just mounting that file by replacing it with yours. You can use the command I mentioned to check it for yourself. Just make sure it has the correct permissions.

**Reply by return42** [reply]
> You can mount the file as a volume and give the address of where you keep your custom logo. for example: /srv/searxng/searxng-custom.png:/usr/local/searxng/searx/static/themes/simple/img/searxng.png

@hirad1986 How do you update SearXNG?

**Reply by hirad1986** [reply]
> @hirad1986 How do you update SearXNG?

Normally! Its just like mounting /srv/searxng:/etc/searxng. The only difference is that the volume is not a directory, but a single (searxng.png) file.

**Reply by return42** [reply]
@hirad1986 thanks .. and this mount works when you upgrade to a new version?

**Reply by hirad1986** [reply]
Yeah. As long as the location of the searxng.png in the source/docker image/container doesn't change, it works.

**Reply by return42** [reply]
OK, thanks .. I think it is OK for now, but there might comes the day the logo will be moved to a different location .. but its not planned. What I know today: The folder structure of SearXNG needs an update, e.g. here https://github.com/searxng/searxng/pull/1188 .. but this PR will not touch the /static path.

**Reply by hirad1986** [reply]
if someday the /static path (and img in particular) changes, the only "issue" that would happen is that the original searxng logo would be displayed, until changing the volume path to the new one.

**Reply by shiftyourname** [reply]
Hey @hirad1986 im trying to follow your footsteps but im a bit lost.
I have added the picture into the docker container under /usr/local/searxng/seaxng-custom.png

But im getting a bit stuck on mounting the /usr/local/searxng/searx/static/themes/simple/img/searxng.png and then give my address instead.

Am i misunderstanding something? i would really love to customize this part of the engine since i will share it with some friends.

Thanks in advance!

**Reply by hirad1986** [reply]
@shiftyourname the file should be on your host machine. Not inside the container. 
When you mount a file or directory, it’s [Path on host]:[Path in container] 
So let’s say you place the file in /srv/searxng/searxng-custom,png. You’d mount the logo this way:
`/srv/searxng/searxng-custom.png:/usr/local/searxng/searx/static/themes/simple/img/searxng.png`

(You can use the directory where you’ve mounted /etc/searxng)

**Reply by shiftyourname** [reply]
Thanks for the reply @hirad1986 <3
So i have made the same directory srv/searxng outside the container with the same file name png in the directory but if i run the command it still says file not found. I have tried running the command in the directory /srv/searxng/ but also inside the container: /usr/local/searxng/searx/static/themes/simple/img/

**Reply by return42** [reply]
Please note, this is just a *searxng-docker* hack .. here we are in the SearXNG repo and the official way to customize SearXNG is to create a fork.

**Reply by hirad1986** [reply]
I'm truly surprised @return42 and not in the good way.
That's not a hack. Forking and modifying it is a hack. Mounting is the official method of customizing a container without making any permanent and/or risky change.
I'm surprised cuz I'm guessing you already know that's what official SearXNG docker does with /etc/searxng, for editing the configs.

**Reply by hirad1986** [reply]
What command are you running @shiftyourname ?!
Are you using docker compose? 
If so, I'm guessing you also mounted/etc/searxng file? Assuming you have and it works just fine, you only need to do the same with this file. Maybe I'll post a docker compose.yml file in my account later as an example, when I have the time.

**Reply by return42** [reply]
> That's not a hack. Forking and modifying it is a hack. 
> Mounting is the official method of customizing a container without making any permanent and/or risky change.

Hi @hirad1986 the official searxng-docker image is one method to [install SearXNG](https://docs.searxng.org/admin/installation.html).  We also offer official alternatives to install without docker (e.g. [Installation Script](https://docs.searxng.org/admin/installation-scripts.html#installation-scripts)).

Docker hacks (quirks, modifications ..) are better discussed in the official https://github.com/searxng/searxng-docker forum.

This discussion is placed in the core SearXNG repo and I'm are core developer and the concept to customize is to create a fork (aka [*brand*](https://docs.searxng.org/admin/engines/settings.html#settings-brand)).

I'm sorry that we have not yet documented how to maintain a *brand* and keep it up-to-date  with the origin of SearXNG ... @mrpaulblack and me committed to do this ASAP.

ATM we have only this documentation about *brand*: https://docs.searxng.org/dev/makefile.html#make-buildenv

In this context you might understand what I mean by:

> TBH we do not have options to customize UI .. if someone wants to customize "this or that" it is recommended to create a fork (we call it a brand) and modify the logo, theme or .. to your needs. Don't forget to rebase your customized branch regularly to get latest updates.

**Comment by return42** [comment, marked as answer]
TBH we do not have options to customize UI .. if someone wants to customize "this or that" it is recommended to create a fork (we call it a *brand*) and modify the logo, theme or .. to your needs.  Don't forget to rebase your customized branch regularly to get latest updates.

**Reply by Exekutori** [reply]
Hm, yes that's a good tip, however in what sense "rebase your customized branch"?

**Reply by return42** [reply]
To get last updates (bugfix) from SearXNG's master branch you can use  **git rebase** ..

- [What is git rebase?](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase)

**Reply by ghost** [reply]
Understanding perfectly @return42 point of view there's also many readers passing by when questionning web search with not enough knowledge of everything mentionned above, ended up not being able to do this changes, branding is at the very bottom and I love it.
But how satisfying is to see your own image when opening your own searxng ? Let me know and please for some using searxng with docker (**not searxng-docker which is pretty easy to be done seeing all folders and files**).
First of all you may get your container id with
`docker ps`
This container id looks like a number so let's say 02d875d7a209 for the steps but you will do with your own container id
According to [docker documentation](https://docs.docker.com/engine/reference/commandline/cp/) and assuming you put your searxng-custom.png and searxng-custom.svg into your /home/user/ please follow instructions below.
`docker cp ./searxng-custom.png 02d875d7a209:/usr/local/searxng/searx/static/themes/simple/img/searxng.png`
`docker cp ./searxng-custom.svg 02d875d7a209:/usr/local/searxng/searx/static/themes/simple/img/searxng.svg`
At this point you will want to restart docker quickly.
`sudo systemctl restart docker`

**Reply by return42** [reply]
>> TBH we do not have options to customize UI 

> But how satisfying is to see your own image when opening your own searxng ?

I still think it will be better we have an option to customize the logo / like other application have it .. we are open for a PR that add such an option / should not hard to implement.
Since I have no need for customizing myself and I also had no time to implement an option, it unfortunately does not yet exist ... but I would be happy if I could review a PR for it .. if anyone has time and interest, please send a PR :+1:

**Reply by ghost** [reply]
>  `docker cp ./searxng-custom.png 02d875d7a209:/usr/local/searxng/searx/static/themes/simple/img/searxng.png` `docker cp ./searxng-custom.svg 02d875d7a209:/usr/local/searxng/searx/static/themes/simple/img/searxng.svg` At this point you will want to restart docker quickly. `sudo systemctl restart docker`

Hi,tnx so much for this,any way to make it permanent? It works and its super easy to apply,but unfortenly default logo is back after container update.....

**Comment by return42** [comment]
> but unfortunately default logo is back after container update.....

logo, favicon, banner .. we are still waiting for PRs for these topics / see feature request:

- https://github.com/searxng/searxng/issues/3208
- https://github.com/searxng/searxng/issues/3176

Instead of every maintainer investing a lot of time in workarounds, wouldn't it be better if someone took the time to implement a PR for the feature?  .. at least that's my opinion ;-)
